### PR TITLE
Forces a side documentation index order

### PR DESCRIPTION
**Pull request by hidmic**
This pull requests enables user defined ordering of the side index. Reproducing the commit message:

_Deepest pages go last. Pages at the same depth are ordered so as to follow rosindex.yml index_pattern order, falling back to alphabetical order to disambiguate._

As an example, with the following `rosindex.yml` configuration file:

```yaml
# The root location to look at for ReST documentation pages.
sources_dir: source

# Patterns for page indexing. Unindexed pages will
# nonetheless be present in rosindex. Useful to keep
# redirecting pages around to guarantee link-wise
# backwards compatibility.
index_pattern:
  - index.rst
  - Concepts.rst
  - Features.rst
  - Roadmap.rst
  - Releases.rst
  - Installation.rst
  - Contributing.rst
  - Tutorials.rst
  - Related-Projects.rst
  - Contact.rst
  - About.rst
  - '**/*.rst'
```

the side index looks like this:

rosindex_sideindex

Note that (1) pages immediately below the root page are ordered following the pattern list order and (2) the rest of the pages are ordered alphabetically, but neither case affects the pages' hierarchy. It cannot.

Builds on top of #139 to avoid rebases.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Can you describe how the resulting nav will look like with this patch. And also give an example for `index_pattern`.

Just from reading the description it sounds that the depth has a higher ranking than the user provided order? If that is the case it would prevent ordering items with different depth in a custom way.

**Comment by hidmic** [comment]
@dirk-thomas I've updated the description. Also, almost by chance, I found and a fixed a bug while doing so. I was sorting page permalinks alphabetically instead of page titles. PTAL now.

**Comment by dirk-thomas** [comment]
> Note that (1) pages immediately below the root page are ordered following the pattern list order and (2) the rest of the pages are ordered alphabetically

How can the order of the items under "ROS 2 Concepts" be influenced? What is the purpose of the last item in index_pattern: `'**/*.rst'`?

**Comment by hidmic** [comment]
> How can the order of the items under "ROS 2 Concepts" be influenced?

You could do:

```yml
index_pattern:
  - index.rst
  - Concepts.rst
  - Concepts/Overview-of-ROS-2-Concepts.rst
  - Concepts/ROS-2-Client-Libraries.rst
  - Concepts/About-ROS-Interfaces.rst
  - Concepts/About-Quality-of-Service-Settings.rst
  - Concepts/DDS-and-ROS-middleware-implementations.rst
  - Concepts/Logging.rst
  - ...
```

> What is the purpose of the last item in index_pattern: `'**/*.rst'`?

The `index_pattern` configuration value provides a list glob expressions that pages' paths relative to the Sphinx source directory have to match to be included on the side index. The last item thus makes sure that all pages that are at least one directory deep show up.

**Comment by dirk-thomas** [comment]
E.g. if I want `Tutorials` to appear higher in the list I can put it into `index_pattern` explicitly at any position. Can I still avoid having to enumerate all tutorial sub pages explicitly? If no, the manual order is only useful when fully enumerating all the pages. If yes, how would the last glob avoid generating duplicates of the `Tutorials/*.rst` pages?

**Comment by hidmic** [comment]
You may have the same page matching multiple patterns, that's not an issue. The logic for inclusion only checks whether _any_ pattern matches the page source path, but won't include it multiple times or anything. 

The logic for ordering (now I see that the commit message was confusing, it's giving too much information) only applies to pages that are at the same hierarchy level i.e. you don't compare pages that are in different directories when sorting. There, if two pages match with two different patterns, the one that matches earlier is the one that goes first. Otherwise, i.e. if two pages match with the same pattern, the order is alphabetical by page title.

**Comment by dirk-thomas** [comment]
Thanks for the explanation. Sounds good to me.

**Comment by hidmic** [comment]
PR rebased and ready for another round (or direct approval :)).

File: _plugins/docs_generator.rb [hunk 1]
```diff
@@ -141,6 +142,28 @@ def convert_with_sphinx(repo_name, repo_path, repo_data)
...
+        end || -1
+        if first_order == second_order
+          first_title = a[1]['title'] || ''
+          second_title = b[1]['title'] || ''
+          first_title <=> second_title
+        else
+          first_order <=> second_order
+        end
+      else
+        first_depth <=> second_depth
```
Comment (dirk-thomas) [inline _plugins/docs_generator.rb:165]: I am wondering why the depth of a navigation item should have any factor in determining the order? That sounds like a pretty arbitrary assumption. E.g if there are sibling items "Foo" and "Bar" I would expect them to be ordered alphabetically and not differently just because "Bar" has a deeper nesting. If in the future additional subpages are created which increase the depth that would also imply that all of a sudden the position of the item in the navigation changes.

File: _plugins/docs_generator.rb [hunk 2]
```diff
@@ -25,7 +25,7 @@ def generate(site)
     site.config['docs_repos'].each do |repo_name, repo_options|
       next unless all_repos.key? repo_name
 
-      repo_path = Pathname.new(File.join("_remotes", repo_name))
+      repo_path = Pathname.new(File.join('_remotes', repo_name))
```
Comment (dirk-thomas) [inline _plugins/docs_generator.rb:28]: I am not sure what style is being used for Ruby but this change looks unrelated and the file mixes both single and double quotes already.

Comment (hidmic) [reply _plugins/docs_generator.rb:28]: It is unrelated, but my OCD triggered and I had to make it consistent at least within the same function i.e. do not use double quotes unless interpolating or printing :).

File: _plugins/docs_generator.rb (hunk 1 above)
Comment (hidmic) [reply _plugins/docs_generator.rb:165]: >if there are sibling items "Foo" and "Bar" I would expect them to be ordered alphabetically and not differently just because "Bar" has a deeper nesting

That's not what that `depth` means in this context. It is the depth of the page w.r.t. the project source root, not the depth of the subtree rooted at said page. For instance, `foo` and `bar` will have the same depth if they are sibling pages. 

As to why that ordering, it is to make it easier for calling code to infer a hierarchy, but it has no impact on how the index is rendered.

**Comment by hidmic** [comment]
@dirk-thomas mind to approve? Or maybe @chapulina or @tfoote can take a look too.

**Comment by dirk-thomas** [comment]
I simply don't know enough about how all this logic works to provide a reasonable review of the patch. If it addresses by feature request (and doesn't break other stuff) I am of course :+1:

**Comment by hidmic** [comment]
Yes, I know, but I don't feel comfortable self-approving/merging. @nuclearsandwich could take a look maybe? I know you know Ruby :). Otherwise, it's fine, I'll merge it tomorrow morning.

**Review by nuclearsandwich — APPROVED** [review]
I approved #143 which includes these changes.

**Comment by hidmic** [comment]
Alright, merging!
